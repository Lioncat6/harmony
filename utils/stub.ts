import { sanitizeFilename } from './sanitize.ts';
import { dirname } from 'std/path/dirname.ts';
import { join } from 'std/path/join.ts';
import { stub } from '@std/testing/mock';

/**
 * Stubs {@linkcode fetch} to load the response from a cache instead of making a network request.
 *
 * Only in download mode (`--download`) the resource will be fetched and written to the cache (as a file).
 */
export function stubFetchWithCache(cacheDir = 'testdata') {
	const originalFetch = globalThis.fetch;
	const downloadMode = Deno.args.includes('--download');

	return stub(globalThis, 'fetch', async function (input, init) {
		const url = new URL(input instanceof Request ? input.url : input);
		const pathSegments = url.href.split('/').map((segment) => sanitizeFilename(segment, '!'));
		const path = join(cacheDir, ...pathSegments);

		if (downloadMode) {
			const response = await originalFetch(input, init);
			const { body } = response.clone();
			if (response.ok && body) {
				await Deno.mkdir(dirname(path), { recursive: true });
				await Deno.writeFile(path, body);
			}
			return response;
		} else {
			try {
				const body = await Deno.readFile(path);
				return new Response(body);
			} catch (error) {
				if (error instanceof Deno.errors.NotFound) {
					throw new Error(`Response for ${url} has not been cached`);
				} else {
					throw error;
				}
			}
		}
	});
}
