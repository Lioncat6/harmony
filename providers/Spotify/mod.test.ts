// Automatically load .env environment variable file (before anything else).
import 'std/dotenv/load.ts';

import { describeProvider, makeProviderOptions } from '@/providers/test_spec.ts';
import { stubFetchWithCache } from '@/utils/stub.ts';
import { describe } from '@std/testing/bdd';

import SpotifyProvider from './mod.ts';

describe('Deezer provider', () => {
	using _fetchStub = stubFetchWithCache();
	const spotify = new SpotifyProvider(makeProviderOptions());

	describeProvider(spotify, {
		urls: [{
			description: 'release page',
			url: new URL('https://open.spotify.com/album/2ZYLme9VXn1Eo8JDtfN7Y9'),
			id: { type: 'album', id: '2ZYLme9VXn1Eo8JDtfN7Y9' },
			isCanonical: true,
		}, {
			description: 'localized release page',
			url: new URL('https://open.spotify.com/intl-de/album/0m8KmmObidPRHOP5knBkam'),
			id: { type: 'album', id: '0m8KmmObidPRHOP5knBkam' },
		}, {
			description: 'artist page',
			url: new URL('https://open.spotify.com/artist/2lD1D6eEh7xQdBtnl2Ik7Y'),
			id: { type: 'artist', id: '2lD1D6eEh7xQdBtnl2Ik7Y' },
			isCanonical: true,
		}, {
			description: 'artist page with tracking parameter',
			url: new URL('https://open.spotify.com/artist/163tK9Wjr9P9DmM0AVK7lm?si=dcd20f1a6b8b49a3'),
			id: { type: 'artist', id: '163tK9Wjr9P9DmM0AVK7lm' },
		}, {
			description: 'track page',
			url: new URL('https://open.spotify.com/track/1EDPVGbyPKJPeGqATwXZvN'),
		}],
		releaseLookup: [],
	});
});
