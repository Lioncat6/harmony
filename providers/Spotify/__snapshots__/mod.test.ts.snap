export const snapshot = {};

snapshot[`Spotify provider > release lookup > single by two artists 1`] = `
{
  artists: [
    {
      creditedName: "Lady Gaga",
      externalIds: [
        {
          id: "1HY2Jd0NmPuamShAr6KMms",
          provider: "spotify",
          type: "artist",
        },
      ],
      name: "Lady Gaga",
    },
    {
      creditedName: "Bruno Mars",
      externalIds: [
        {
          id: "0du5cEVh5yTK9QJze8zA0C",
          provider: "spotify",
          type: "artist",
        },
      ],
      name: "Bruno Mars",
    },
  ],
  availableIn: [
    "AR",
    "AU",
    "AT",
    "BE",
    "BO",
    "BR",
    "BG",
    "CA",
    "CL",
    "CO",
    "CR",
    "CY",
    "CZ",
    "DK",
    "DO",
    "DE",
    "EC",
    "EE",
    "SV",
    "FI",
    "FR",
    "GR",
    "GT",
    "HN",
    "HK",
    "HU",
    "IS",
    "IE",
    "IT",
    "LV",
    "LT",
    "LU",
    "MY",
    "MT",
    "MX",
    "NL",
    "NZ",
    "NI",
    "NO",
    "PA",
    "PY",
    "PE",
    "PH",
    "PL",
    "PT",
    "SG",
    "SK",
    "ES",
    "SE",
    "CH",
    "TW",
    "TR",
    "UY",
    "US",
    "GB",
    "AD",
    "LI",
    "MC",
    "ID",
    "JP",
    "TH",
    "VN",
    "RO",
    "IL",
    "ZA",
    "SA",
    "AE",
    "BH",
    "QA",
    "OM",
    "KW",
    "EG",
    "MA",
    "DZ",
    "TN",
    "LB",
    "JO",
    "PS",
    "IN",
    "KZ",
    "MD",
    "UA",
    "AL",
    "BA",
    "HR",
    "ME",
    "MK",
    "RS",
    "SI",
    "KR",
    "BD",
    "PK",
    "LK",
    "GH",
    "KE",
    "NG",
    "TZ",
    "UG",
    "AG",
    "AM",
    "BS",
    "BB",
    "BZ",
    "BT",
    "BW",
    "BF",
    "CV",
    "CW",
    "DM",
    "FJ",
    "GM",
    "GE",
    "GD",
    "GW",
    "GY",
    "HT",
    "JM",
    "KI",
    "LS",
    "LR",
    "MW",
    "MV",
    "ML",
    "MH",
    "FM",
    "NA",
    "NR",
    "NE",
    "PW",
    "PG",
    "WS",
    "SM",
    "ST",
    "SN",
    "SC",
    "SL",
    "SB",
    "KN",
    "LC",
    "VC",
    "SR",
    "TL",
    "TO",
    "TT",
    "TV",
    "VU",
    "AZ",
    "BN",
    "BI",
    "KH",
    "CM",
    "TD",
    "KM",
    "GQ",
    "SZ",
    "GA",
    "GN",
    "KG",
    "LA",
    "MO",
    "MR",
    "MN",
    "NP",
    "RW",
    "TG",
    "UZ",
    "ZW",
    "BJ",
    "MG",
    "MU",
    "MZ",
    "AO",
    "CI",
    "DJ",
    "ZM",
    "CD",
    "CG",
    "IQ",
    "LY",
    "TJ",
    "VE",
    "ET",
    "XK",
  ],
  copyright: "© 2024 Interscope Records
℗ 2024 Interscope Records",
  excludedFrom: [
    "BY",
    "PR",
  ],
  externalLinks: [
    {
      types: [
        "free streaming",
      ],
      url: "https://open.spotify.com/album/10FLjwfpbxLmW8c25Xyc2N",
    },
  ],
  gtin: "00602475093060",
  images: [
    {
      thumbUrl: "https://i.scdn.co/image/ab67616d00001e0282ea2e9e1858aa012c57cd45",
      types: [
        "front",
      ],
      url: "https://i.scdn.co/image/ab67616d0000b27382ea2e9e1858aa012c57cd45",
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: "https://api.spotify.com/v1/albums/10FLjwfpbxLmW8c25Xyc2N",
        id: "10FLjwfpbxLmW8c25Xyc2N",
        internalName: "spotify",
        lookup: {
          method: "id",
          value: "10FLjwfpbxLmW8c25Xyc2N",
        },
        name: "Spotify",
        url: "https://open.spotify.com/album/10FLjwfpbxLmW8c25Xyc2N",
      },
    ],
  },
  labels: [
    {
      name: "Interscope",
    },
  ],
  media: [
    {
      format: "Digital Media",
      number: 1,
      tracklist: [
        {
          artists: [
            {
              creditedName: "Lady Gaga",
              externalIds: [
                {
                  id: "1HY2Jd0NmPuamShAr6KMms",
                  provider: "spotify",
                  type: "artist",
                },
              ],
              name: "Lady Gaga",
            },
            {
              creditedName: "Bruno Mars",
              externalIds: [
                {
                  id: "0du5cEVh5yTK9QJze8zA0C",
                  provider: "spotify",
                  type: "artist",
                },
              ],
              name: "Bruno Mars",
            },
          ],
          availableIn: [
            "AR",
            "AU",
            "AT",
            "BE",
            "BO",
            "BR",
            "BG",
            "CA",
            "CL",
            "CO",
            "CR",
            "CY",
            "CZ",
            "DK",
            "DO",
            "DE",
            "EC",
            "EE",
            "SV",
            "FI",
            "FR",
            "GR",
            "GT",
            "HN",
            "HK",
            "HU",
            "IS",
            "IE",
            "IT",
            "LV",
            "LT",
            "LU",
            "MY",
            "MT",
            "MX",
            "NL",
            "NZ",
            "NI",
            "NO",
            "PA",
            "PY",
            "PE",
            "PH",
            "PL",
            "PT",
            "SG",
            "SK",
            "ES",
            "SE",
            "CH",
            "TW",
            "TR",
            "UY",
            "US",
            "GB",
            "AD",
            "LI",
            "MC",
            "ID",
            "JP",
            "TH",
            "VN",
            "RO",
            "IL",
            "ZA",
            "SA",
            "AE",
            "BH",
            "QA",
            "OM",
            "KW",
            "EG",
            "MA",
            "DZ",
            "TN",
            "LB",
            "JO",
            "PS",
            "IN",
            "KZ",
            "MD",
            "UA",
            "AL",
            "BA",
            "HR",
            "ME",
            "MK",
            "RS",
            "SI",
            "KR",
            "BD",
            "PK",
            "LK",
            "GH",
            "KE",
            "NG",
            "TZ",
            "UG",
            "AG",
            "AM",
            "BS",
            "BB",
            "BZ",
            "BT",
            "BW",
            "BF",
            "CV",
            "CW",
            "DM",
            "FJ",
            "GM",
            "GE",
            "GD",
            "GW",
            "GY",
            "HT",
            "JM",
            "KI",
            "LS",
            "LR",
            "MW",
            "MV",
            "ML",
            "MH",
            "FM",
            "NA",
            "NR",
            "NE",
            "PW",
            "PG",
            "WS",
            "SM",
            "ST",
            "SN",
            "SC",
            "SL",
            "SB",
            "KN",
            "LC",
            "VC",
            "SR",
            "TL",
            "TO",
            "TT",
            "TV",
            "VU",
            "AZ",
            "BN",
            "BI",
            "KH",
            "CM",
            "TD",
            "KM",
            "GQ",
            "SZ",
            "GA",
            "GN",
            "KG",
            "LA",
            "MO",
            "MR",
            "MN",
            "NP",
            "RW",
            "TG",
            "UZ",
            "ZW",
            "BJ",
            "MG",
            "MU",
            "MZ",
            "AO",
            "CI",
            "DJ",
            "ZM",
            "CD",
            "CG",
            "IQ",
            "LY",
            "TJ",
            "VE",
            "ET",
            "XK",
          ],
          isrc: "USUM72409273",
          length: 251667,
          number: 1,
          recording: {
            externalIds: [
              {
                id: "2plbrEY59IikOBgBGLjaoe",
                provider: "spotify",
                type: "track",
              },
            ],
          },
          title: "Die With A Smile",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 16,
    month: 8,
    year: 2024,
  },
  status: "Official",
  title: "Die With A Smile",
  types: [
    "Single",
  ],
}
`;
